File: ios/core/Sources/Types/Core/ErrorController.swift
//
//  ErrorController.swift
//  PlayerUI
//
//  Created by Player Team
//

import Foundation
import JavaScriptCore

/// Severity levels for errors
public enum ErrorSeverity: String {
    /// Cannot continue, flow must end
    case fatal
    /// Standard error, may allow recovery
    case error
    /// Non-blocking, logged for telemetry
    case warning
}

/// Known error types for Player
public enum ErrorTypes: Equatable {
    case expression
    case binding
    case view
    case asset
    case navigation
    case validation
    case data
    case schema
    case network
    case plugin
    case render
    /// An error type not recognized by this version of the SDK.
    case unknown(String)

    public init(_ rawValue: String) {
        switch rawValue {
        case "expression": self = .expression
        case "binding": self = .binding
        case "view": self = .view
        case "asset": self = .asset
        case "navigation": self = .navigation
        case "validation": self = .validation
        case "data": self = .data
        case "schema": self = .schema
        case "network": self = .network
        case "plugin": self = .plugin
        case "render": self = .render
        default: self = .unknown(rawValue)
        }
    }

    public var rawValue: String {
        switch self {
        case .expression: return "expression"
        case .binding: return "binding"
        case .view: return "view"
        case .asset: return "asset"
        case .navigation: return "navigation"
        case .validation: return "validation"
        case .data: return "data"
        case .schema: return "schema"
        case .network: return "network"
        case .plugin: return "plugin"
        case .render: return "render"
        case .unknown(let value): return value
        }
    }
}

/// A wrapper around the JS ErrorController in the core player
public class ErrorController: CreatedFromJSValue {
    /// Typealias for associated type
    public typealias T = ErrorController

    /// Creates an instance from a JSValue, used for generic construction
    /// - Parameters:
    ///   - value: The JSValue to construct from
    public static func createInstance(value: JSValue) -> ErrorController { ErrorController(value) }

    /// The JSValue that backs this wrapper
    private let value: JSValue

    /// The hooks that can be tapped into
    public let hooks: ErrorControllerHooks

    /// Construct an ErrorController from a JSValue
    /// - Parameters:
    ///   - value: The JSValue that is the ErrorController
    public init(_ value: JSValue) {
        self.value = value
        hooks = ErrorControllerHooks(
            onError: HookWithResult<JSValueError, Bool>(baseValue: value, name: "onError")
        )
    }

    /// Capture an error
    /// - Parameters:
    ///   - error: The native Error object
    /// - Returns: Whether the error was successfully captured
    @discardableResult
    public func captureError(
        error: Error
    ) -> Bool {
        let convertibleError = (error as? JSConvertibleError & Error) ?? PlayerError.unknownResponse(error)
        let args = [value.context.error(for: convertibleError) as Any]

        guard let result = value.invokeMethod("captureError", withArguments: args), result.isBoolean else {
            return false
        }
        return result.toBool()
    }

    /// Get the most recent error
    /// - Returns: The current error as a JSValue if one exists
    public func getCurrentError() -> JSValue? {
        return value.invokeMethod("getCurrentError", withArguments: [])
    }

    /// Get the complete error history
    /// - Returns: JSValue representing the array of errors
    public func getErrors() -> JSValue? {
        return value.invokeMethod("getErrors", withArguments: [])
    }

    /// Clear all errors (history + current + data model)
    public func clearErrors() {
        value.invokeMethod("clearErrors", withArguments: [])
    }

    /// Clear only current error and remove from data model, preserve history
    public func clearCurrentError() {
        value.invokeMethod("clearCurrentError", withArguments: [])
    }
}

File: jvm/core/src/main/kotlin/com/intuit/playerui/core/error/ErrorController.kt
package com.intuit.playerui.core.error

import com.intuit.playerui.core.bridge.Node
import com.intuit.playerui.core.bridge.NodeWrapper
import com.intuit.playerui.core.bridge.hooks.NodeSyncBailHook1
import com.intuit.playerui.core.bridge.serialization.serializers.NodeSerializableField
import com.intuit.playerui.core.bridge.serialization.serializers.NodeSerializableFunction
import com.intuit.playerui.core.bridge.serialization.serializers.NodeWrapperSerializer
import com.intuit.playerui.core.bridge.serialization.serializers.ThrowableSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.serializer

/** Severity levels for errors */
@Serializable
public enum class ErrorSeverity(
    public val value: String,
) {
    /** Cannot continue, flow must end */
    @SerialName("fatal")
    FATAL("fatal"),

    /** Standard error, may allow recovery */
    @SerialName("error")
    ERROR("error"),

    /** Non-blocking, logged for telemetry */
    @SerialName("warning")
    WARNING("warning"),
}

/** Known error types for Player */
@Serializable
public enum class ErrorTypes(
    public val value: String,
) {
    @SerialName("expression")
    EXPRESSION("expression"),

    @SerialName("binding")
    BINDING("binding"),

    @SerialName("view")
    VIEW("view"),

    @SerialName("asset")
    ASSET("asset"),

    @SerialName("navigation")
    NAVIGATION("navigation"),

    @SerialName("validation")
    VALIDATION("validation"),

    @SerialName("data")
    DATA("data"),

    @SerialName("schema")
    SCHEMA("schema"),

    @SerialName("network")
    NETWORK("network"),

    @SerialName("plugin")
    PLUGIN("plugin"),

    @SerialName("render")
    RENDER("render"),
}

/**
 * Limited definition of the player error controller to enable error capture and management
 */
@Serializable(with = ErrorController.Serializer::class)
public class ErrorController internal constructor(
    override val node: Node,
) : NodeWrapper {
    private val captureError: (error: Throwable) -> Boolean by NodeSerializableFunction<Boolean>()
    private val getCurrentError: () -> Throwable? by NodeSerializableFunction<Throwable?>()
    private val getErrors: () -> List<Throwable>? by NodeSerializableFunction<List<Throwable>?>()
    private val clearErrors: () -> Unit by NodeSerializableFunction<Unit>()
    private val clearCurrentError: () -> Unit by NodeSerializableFunction<Unit>()

    public val hooks: Hooks by NodeSerializableField(Hooks.serializer())

    /**
     * Capture an error
     * @param error The error/exception object
     * @return Whether the error was handled
     */
    public fun captureError(error: Throwable): Boolean = captureError.invoke(error)

    /**
     * Get the most recent error
     * @return The current error if one exists
     */
    public fun getCurrentError(): Throwable? = getCurrentError.invoke()

    /**
     * Get the complete error history
     * @return List of all captured errors in chronological order
     */
    public fun getErrors(): List<Throwable>? = getErrors.invoke()

    /**
     * Clear all errors (history + current + data model)
     */
    public fun clearErrors() {
        clearErrors.invoke()
    }

    /**
     * Clear only current error and remove from data model, preserve history
     */
    public fun clearCurrentError() {
        clearCurrentError.invoke()
    }

    @Serializable(Hooks.Serializer::class)
    public class Hooks internal constructor(
        override val node: Node,
    ) : NodeWrapper {
        /**
         * Fired when any error is captured
         * - The callback receives a PlayerErrorInfo object
         * - Return true from the callback to bail and prevent error state navigation
         * - Return false/null to continue to next handler
         */
        public val onError: NodeSyncBailHook1<Throwable, Boolean>
            by NodeSerializableField(NodeSyncBailHook1.serializer(ThrowableSerializer(), Boolean.serializer()))

        internal object Serializer : NodeWrapperSerializer<Hooks>(::Hooks)
    }

    internal object Serializer : NodeWrapperSerializer<ErrorController>(::ErrorController)
}
